### ISSUE TITLE
Horrible socket handling leads to dozens of UI and state bugs

### ISSUE THREAD

**Issue by MatthewRalston** [galaxyproject/galaxy#2320]
```
----------------------------------------
Exception happened during processing of request from ('1.2.3.4', 38112)
Traceback (most recent call last):
  File "/apps/sys/galaxy/galaxy_env/lib/python2.7/site-packages/paste/httpserver.py", line 1068, in process_request_in_thread
    self.finish_request(request, client_address)
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 653, in __init__
    self.finish()
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 712, in finish
    self.wfile.close()
  File "/apps/sys/python/2.7.8/lib/python2.7/socket.py", line 279, in close
    self.flush()
  File "/apps/sys/python/2.7.8/lib/python2.7/socket.py", line 303, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
error: [Errno 32] Broken pipe
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('1.2.3.4', 38119)
Traceback (most recent call last):
  File "/apps/sys/galaxy/galaxy_env/lib/python2.7/site-packages/paste/httpserver.py", line 1068, in process_request_in_thread
    self.finish_request(request, client_address)
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 653, in __init__
    self.finish()
  File "/apps/sys/python/2.7.8/lib/python2.7/SocketServer.py", line 712, in finish
    self.wfile.close()
  File "/apps/sys/python/2.7.8/lib/python2.7/socket.py", line 279, in close
    self.flush()
  File "/apps/sys/python/2.7.8/lib/python2.7/socket.py", line 303, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
error: [Errno 32] Broken pipe
----------------------------------------
galaxy.jobs.runners.local DEBUG 2016-05-05 16:45:17,145 (12347) executing job script: /apps/sys/galaxy/galaxy-v15.05-production/database/job_working_directory/012/12347/galaxy_12347.sh
galaxy.jobs DEBUG 2016-05-05 16:45:17,463 (12347) Persisting job destination (destination id: local:///)
galaxy.jobs WARNING 2016-05-05 16:45:17,745 (12347) Ignoring state change from 'error' to 'running' for job that is already terminal
galaxy.jobs.runners.local DEBUG 2016-05-05 16:45:23,436 execution finished: /apps/sys/galaxy/galaxy-v15.05-production/database/job_working_directory/012/12347/galaxy_12347.sh

```

The errors take the form above. This has lead to datasets being greyed out, failing to begin execution, and it has also lead to the complete failure to rerun a job stopped by an administrator. A colleague tried to import a dataset (which leads to the "upload" script being run by an anonymous user instead of the current user). As an administrator, I stopped the job in panic, as anonymous jobs are not even enabled on our server. After realizing that this is how Galaxy imports datasets from libraries, I impersonated the user, reran the import, which then failed with the same Administrative message that I originally canceled the job with. I have restarted the galaxy instance, deleted the failed imports (which ironically have all of the data in them when clicking on the view button), and reran all with the same end result: a completely unusable instance of Galaxy. </rant>

In summary:
when sockets fail they make galaxy virtually unusable, and trigger all sorts of state malfunctions such as:
- jobs perpetually stuck in the queue
- jobs being rerun after being set to error
- successful jobs are marked as errored
- datasets from jobs incorrectly marked as errored or queued containing full and correct data.
- Future jobs on these complete and incorrectly labeled datasets cannot be run, as the state is marked incorrectly in the database.

**Comment by MatthewRalston** [comment]
![galaxy-fail](https://cloud.githubusercontent.com/assets/4308024/15057712/339b5f32-12e6-11e6-89f6-23887bc13382.png)

I mistakenly deleted a job where a user was importing some files into their history. Now, whenever I import datasets into their history (after impersonating them) the datasets are marked as errored. However, the exit status is 0 and the dataset contains the correct information (gzip file shows up as symbols here).

![galaxy-fail1](https://cloud.githubusercontent.com/assets/4308024/15057788/a6688ecc-12e6-11e6-8c2c-0d417e14e77c.png)

**Comment by MatthewRalston** [comment]
Related perhaps: https://bugs.python.org/issue14574
